stan/math/rev/mat/functor/kinsol_solve.hpp

```diff
@@ -0,0 +1,126 @@
+#include <stan/math/rev/mat/functor/kinsol_data.hpp>
+#include <stan/math/rev/mat/functor/algebra_system.hpp>
+#include <stan/math/prim/mat/err/check_flag.hpp>
+
+#include <kinsol/kinsol.h>
+#include <sunmatrix/sunmatrix_dense.h>
+#include <sunlinsol/sunlinsol_dense.h>
+#include <nvector/nvector_serial.h>
+
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the solution to the specified algebraic system,
+ * given an initial guess. Invokes the Kinsol solver from Sundials.
+ *
+ * @tparam F type of equation system function.
+ * @tparam T type of initial guess vector.
+ * @param[in] f Functor that evaluated the system of equations.
+ * @param[in] x Vector of starting values.
+ * @param[in] y Parameter vector for the equation system. The function
+ *            is overloaded to treat y as a vector of doubles or of a
+ *            a template type T.
+ * @param[in] dat Continuous data vector for the equation system.
+ * @param[in] dat_int Integer data vector for the equation system.
+ * @param[in, out] msgs The print stream for warning messages.
+ * @param[in] scaling_step_tol Scaled-step stopping tolerance. If
+ *            a Newton step is smaller than the scaling step
+ *            tolerance, the code breaks, assuming the solver is no
+ *            longer making significant progress (i.e. is stuck)
+ * @param[in] function_tolerance determines whether roots are acceptable.
+ * @param[in] max_num_steps Maximum number of function evaluations.
+ * @param[in] custom_jacobian If 0, use Kinsol's default to compute the
+ *            jacobian for the Newton step, namely Quotient Difference
+ *            (finite difference). If 1, use reverse-mode AD, unless
+ *            the user specifies their own method.
+ * @param[in] J_f A functor that computes the Jacobian for the Newton step.
+ *            Defaults to reverse-mode AD.
+ * @param[in] steps_eval_jacobian Maximum number of steps before the
+ *            Jacobian gets recomputed. Note that Kinsol's default is 10.
+ *            If equal to 1, the algorithm computes exact Newton steps.
+ * @param[in] global_line_search does the solver use a global line search?
+ *            If equal to KIN_NONE, no, if KIN_LINESEARCH, yes.
+ * @return x_solution Vector of solutions to the system of equations.
+ * @throw <code>std::invalid_argument</code> if Kinsol returns a negative
+ *        flag when setting up the solver.
+ * @throw <code>boost::math::evaluation_error</code> if Kinsol returns a
+ *        negative flag after attempting to solve the equation.
+ */
+template <typename F1, typename F2 = kinsol_J_f>
+Eigen::VectorXd kinsol_solve(
+    const F1& f, const Eigen::VectorXd& x, const Eigen::VectorXd& y,
+    const std::vector<double>& dat, const std::vector<int>& dat_int,
+    std::ostream* msgs = nullptr, double scaling_step_tol = 1e-3,
+    double function_tolerance = 1e-6,
+    long int max_num_steps = 1e+3,  // NOLINT(runtime/int)
```

Comment:
First question, where do you set `max_num_steps`? I looked for `KINSetNumMaxIters` but couldn't find it. Without setting it the later `check_flag_kinsol` would be reporting wrong results.

Second question, why set max iter to 1000 instead of using default 200? A grief I have on ODE solvers is that many control parameters are out of thin air without documented justification.

Reply:
Good catch. The default of 1000 is conservative, but in practice, I've never dealt with systems that take more 20 iterations to solve. I'll set the default to 200.